### ISSUE TITLE
DEV: `--doctests` option to `runtests.py` doesn't do anything

### ISSUE THREAD

**Issue by rossbar** [numpy/numpy#21070]
The `runtests.py` script has a `--doctests` command line option with the following description:

>   --doctests            Run doctests in module

However, AFAICT this always raises a `ValueError`. I tried several invocations: `python runtests.py --doctests`, `python runtests.py --doctests numpy/lib/histograms.py` (in an attempt to limit the tests to a single module). No matter what, this seems to call down to [_pytesttester.py](https://github.com/numpy/numpy/blob/5a53fa9326047e33c251d238a6ce344980cbe038/numpy/_pytesttester.py#L169-L170) which simply raises a `ValueError` saying that doctests are not supported. 

I assume this is just a vestigial option that should be removed from `runtests.py`, but I wanted to check if this was expected to work.

**Comment by rossbar** [comment]
In a similar vein, [this info](https://numpy.org/devdocs/reference/testing.html#doctests) also seems to be out of date [(source)](https://github.com/numpy/numpy/blob/main/doc/TESTS.rst.txt)

**Comment by mattip** [comment]
I would prefer to move to a better supported doctest runner like pytest-doctestplus. See #15845

**Comment by rossbar** [comment]
> I'm not aware of misbehaviors of the refguide-check doctesting functionality, could somebody give me a pointer?

I wouldn't say that it's so much "misbehaviors" as not really having confidence in what `refguide_check` is doing. There are at least 4 separate skiplists - in many cases it's not immediately clear why something is on one list or another, or even what tests are actually being skipped. There was recently a PR (#21064) that fixed a trivial repr/spacing issue for a doctest that was failing in `astropy`'s test suite, but was not being caught in `numpy`'s. The fact that this wasn't caught further erodes confidence that the refguide/doctests are being picked up properly.

> Extracting the refguide-check DTChecker into a pytest plugin was/is on my TODO list for ages, and I still think it makes sense. And I do believe it's worth doing irrespective of the decision on moving over to doctest-plus or not.

Wholeheartedly agree - being able run the refguide checks within the pytest framework would go a long way toward alleviating many of the concerns. For example, just being able to do pytest -v to see which tests are actually running would already be a very nice improvement!

**Comment by rossbar** [comment]
>  As far as I can quickly see, some of these irrelevant to numpy (scipy stuff), some are unrelated to doctesting (checks ReST syntax) and some are probably unavoidable in some form (will migrate to conftest.py or some such: these are either docstrings not to be doctested or which cannot be doctested without e.g. importing from scipy).

Agreed - it could just be that we need to audit these lists and verify that the things that are excluded *should* be.

> If 1 is moved out of the picture (would listen to suggestions), the rest is clearly doable. The main Q is person-power to actually do it.

I'd have to look more closely at just how extensive the rst validation is, but my initial thought is that sphinx+numpydoc *should* be sufficient for catching rst problems during the sphinx build. But I agree that we should verify this... there may even be some rst validation in `refguide_check` that should be upstreamed to `numpydoc`.

> The main Q is person-power to actually do it.

Naturally :)

> Would you be interested / available? I can help out as much as I can, but cannot promise a definite ETA if I'm the main champion on this.

I'm certainly interested, though I'm in a similar boat where I don't think I'd be able to necessarily rocket this to the top of my priority queue... @melissawm had the excellent idea of perhaps pursuing a small development grant for this project (see also: [community meeting archive](https://github.com/numpy/archive/blob/main/community_meetings/community-2022-05-11.md#topics)).

**Comment by ev-br** [comment]
> For example, just being able to do pytest -v to see which tests are actually running would already be a very nice improvement!

There is a `-v` argument to `refguide-check`, but it's broken: running `$ python runtests.py --refguide-check -s core -vvv` still has `args.verbose` being zero (I just dropped `import pdb; pdb.set_trace()` at https://github.com/numpy/numpy/blob/main/tools/refguide_check.py#L1145 ; it is then used at https://github.com/numpy/numpy/blob/main/tools/refguide_check.py#L1198). 

So this could be a first easy-fix issue :-).

**Comment by ev-br** [comment]
>>There was recently a PR (#21064)
>
> These things should indeed be handled by refguide-check automatically, feel free to ping me on anything of this sort --- so far > we've not seen these things in scipy for quite a while.

Actually, I take this back. In fact, the behavior in #21064 is by design: refguide-check working as expected! It's been designed to actually ignore whitespace differences, so the output could have been written as `array([0., 3.14159265, 0.78539816])`. What it does under the hood is essentially

```
In [2]: np.allclose([0, 1, 2], [   0   ,   1,      2  ])
Out[2]: True
```

which I would argue is _quite right_ :-).

**Comment by rossbar** [comment]
> which I would argue is quite right :-).

I agree that this is an improvement over the standard doctest behavior - enforcing exact character matching of outputs has always been a major limitation IMO. However, it wasn't obvious (to me at least, though I never bothered to look at the code) that this is how `refguide_check` works! At the very least, the deviation from "vanilla" doctest behavior should be documented so we can account for differences between `refguide_check` and other methods that still rely on standard doctest behavior, like whatever workflow led to #21064 or sphinx's `doctest` hooks (not that anyone *should* be using those).

**Comment by bsipocz** [comment]
this one example onewhere above is I think an unfixed bug in doctestplus, at least there is a long-standing issue for it: https://github.com/astropy/pytest-doctestplus/issues/92

(and the underlying cause of some of the directives are not working nicely together). That being said I think it would be superb to converge all the tools into one plugin that pulls maintenance resources together, too rather than maintaining several flavours of infrastructure that basically doing the same thing.

**Comment by ev-br** [comment]
I'd argue that it's one case where diversity in the ecosystem is good.

**Comment by bsipocz** [comment]
I would love to learn more about your reasoning for this, especially for maintaining two almost identical, but not exactly identical tools for such close projects as scipy and numpy.

As to whether move into the pytest ecosystem with doctestplus, I fully understand that the separation is largely due to history and that many projects didn't use pytest even up to a few years ago. And I know that changing tools need a bigger buy in, but overall I think the ecosystem would benefit if all the resources that go into the parallel, homegrown solutions would go into one, better maintained one that becomes the de-facto standard in the wider scientific python ecosystem.

**Comment by ev-br** [comment]
In case someone's interested, https://github.com/ev-br/scipydoctest implements a chewing-gum-and-scotch-tape variety of a prototype for a pytest plugin with modified doctesting from refguide-check. Not that it works yet on anything more than a toy example, but at least it no longer needs to be tied to a homegrown test runner/collector, so it might be easier to evolve.

**Comment by ev-br** [comment]
> I would love to learn more about your reasoning for this, especially for maintaining two almost identical, but not exactly identical tools for such close projects as scipy and numpy.

As far as I'm concerned, my reasoning is not much beyond that diversity is better than a monoculture, in general. 

For these tools in general, I think both need work (not that the current scipy solution needed much maintenance over time). For pytest-doctestplus specifically, I guess it is not at all clear how much is needed to actually use it for numpy and scipy (*a lot* of complexity in refguide-check is driven by actual needs of existing body of documentation). For refguide-check based solution, yes, it needs work to transition from a homemade runner to a standard pytest one. As an exercise from the previous comment shows it's certainly doable. The question --- for both tools --- is the last mile.

**Comment by ev-br** [comment]
Now that I poked around the pytest internals (am going to do the same with pytest-doctestplus some time next week), I think the following seem to be deal-breakers for scipy at least, maybe for numpy as well:
- an ability to switch on `+FLOAT_CMP` globally
- an ability to select subsets of objects to doctest. pytest built-in test collection is nice to rely on, (I'm still figuring out how to only select doctests), but what is also needed is an ability to e.g. only select public objects. Roughly, a switch to say "only collect doctests and only from objects which are present in `__all__` listsings of their modules". 
- block lists : an ability to say "these docstrings are not to be doctested"; these should be per-project, I suppose.
- a way to handle namedtuples. Basically, given `MoodResult(statistic=42, pvalue=0.5)`, only compare the tuple `(42, 0.5)`.
- additional markers, `# may vary` et al instead of `# doctest: +SKIP`
- lists of stopwords (MPL commands)
- some flexibility of manipulating what is a single doctest in rst files (was needed for the tutorial, is quite hackish in refguide-check ATM).

I'd love to hear your thoughts on these @bsipocz

**Comment by bsipocz** [comment]
So, this is a full circle today. After triaging doctest related issues, I run into this same issue of not knowing how to run doctests here. I think all the docs should be fixed for now, even if it's just temporary, as promoting the ValueError from `numpy.test(doctests=True)` is more confusing than helpful.

However, apparent`rundocs` seems to be the way to run the doctests: https://github.com/numpy/numpy/pull/21566#issuecomment-1133820090, so I'll spend a bit more time to see how easy to plug this into `_pytesttester`.

**Comment by bsipocz** [comment]
@ev-br - I'll need to do a bit more digging to have the answers for all your points, this is my first recollection:


>     * an ability to switch on `+FLOAT_CMP` globally

Yes, it's possible to configure in setup.cfg: https://github.com/astropy/pytest-doctestplus#floating-point-comparison

>     * an ability to select subsets of objects to doctest. pytest built-in test collection is nice to rely on, (I'm still figuring out how to only select doctests), but what is also needed is an ability to e.g. only select public objects. Roughly, a switch to say "only collect doctests and only from objects which are present in `__all__` listsings of their modules".

I don't know how to do this, but the feature is a reasonable one, but I can't judge how difficult it would be to implement. Curiosity: why do you want to not test private functionality?
(I assume here you want a switch, so you can still run the private ones. As if you never want to run those it's very easy to do either with the function dependent skip directives, or simply by not writing valid docstring examples (e.g. omitting the leading `>>>`)

>     * block lists : an ability to say "these docstrings are not to be doctested"; these should be per-project, I suppose.

There are a lot of flavours of skip directives, skipping lines (as normal doctest), to skipping snippets, to skipping whole files. Also, there are ways to skip if a dependency is not available, or if remote-data is required (this likely is not relevant for you). I think these cover most of the use cases, but new directives could be added if something else is missing.
https://github.com/astropy/pytest-doctestplus#skipping-tests

>     * a way to handle namedtuples. Basically, given `MoodResult(statistic=42, pvalue=0.5)`, only compare the tuple `(42, 0.5)`.

I don't think this is supported atm, and I vaguely recall to have similar issues when doctested examples in a book manuscript. (btw, yes, it supports other file formats, not just rst, but TeX, md, txt too)

>     * additional markers, `# may vary` et al instead of `# doctest: +SKIP`

There are only `# doctest` markers, but there are a wide variety of them. You may like e.g. `IGNORE_OUTPUT` that runs the line but ignores the output (but fails the test if an exception/warning is raised), or `IGNORE_WARNINGS` is very handy, too.
 
>     * lists of stopwords (MPL commands)

there are no stopwords. We deal with mpl command either by skipping them/the snippet, or use `IGNORE_OUTPUT`, or use the plot directive (then it's being tested during docs generation).

>     * some flexibility of manipulating what is a single doctest in rst files (was needed for the tutorial, is quite hackish in refguide-check ATM).

doctests context can be restarted, so it's possible to define multiple ones in a file (as I recall the default is that the whole file is one big test, thus if there is a failure the execution is stopped at that point).

**Comment by bsipocz** [comment]
Btw, there have been ideas floated about upstreaming doctestplus to pytest (https://github.com/astropy/pytest-doctestplus/issues/100), or numpy/scipy (https://github.com/astropy/pytest-doctestplus/issues/55), etc, but none of the ideas was championed so status-quo remained.

And most of the dev work there is champion dependent (I needed Tex support and remote-data, no one stopped me getting them if I wrote/mentored the PR (but of course got reviews on them)), and I can't really recall any feature adding PRs that were turned down, given they are not breaking current functionality.


Caveats that come to mind: 
- new release of pytest regularly breaks doctestplus. I don't think that is doctestplus related, and feel that other plugins would be affected as well.
- doctest likes to have full examples. E.g. import statements, and definitions. e.g. one cannot just say in the middle of a docstring that `np.mean(input_array)` without saying anything about `np` and `input_array`. But I think this is also a case with vanilla doctest. 
- `# doctest: +SKIP` and other comments are visible in the docstring from the python interpreter. Which is ugly. But they are rendered out by sphinx so the narrative docs/API docs will look great.
- there are bugs but not enough maintainer time. So most of us fix the ones that bite our own projects. Thus a few, like that whitespace+float bug is still around. But I think this is not a unique situation, and one of the main reasons why I feel it would be great to pull resources together.

**Comment by ev-br** [comment]
>> * an ability to select subsets of objects to doctest. pytest built-in test collection is nice to rely on, (I'm still figuring out how to 
>> only select doctests), but what is also needed is an ability to e.g. only select public objects. Roughly, a switch to say "only 
>> collect doctests and only from objects which are present in `__all__` listsings of their modules".

> I don't know how to do this, but the feature is a reasonable one, but I can't judge how difficult it would be to implement. 
> Curiosity: why do you want to not test private functionality?
> (I assume here you want a switch, so you can still run the private ones. As if you never want to run those it's very easy to do 
> either with the function dependent skip directives, or simply by not writing valid docstring examples (e.g. omitting the leading 
> `>>>`)

That's how the current refguide-check works: it only selects the public objects modules (which are explicitly listed), classes and functions. As to why, well, I think the requirements for private functions need not be as strict as for the user-facing ones.

And indeed, I'm looking for a switch. The current refguide-check is inflexible (doctests whole modules only), and making the best of both worlds would be beneficial IMO.

> There are a lot of flavours of skip directives, skipping lines (as normal doctest), to skipping snippets, to skipping whole files. 
> Also, there are ways to skip if a dependency is not available, or if remote-data is required (this likely is not relevant for you). I > think these cover most of the use cases, but new directives could be added if something else is missing.

My reasoning is that a user does not need to worry whether a docstring is doctested or not, hence a skip list should be kept in a tool. 

> There are only # doctest markers, but there are a wide variety of them. You may like e.g. IGNORE_OUTPUT that runs the line > but ignores the output (but fails the test if an exception/warning is raised), or IGNORE_WARNINGS is very handy, too.

Ditto: doctest markers are line noise, while `>>> np.random.uniform(size=5)   #  may vary` reads like a text. The current refguide-check picks these up, and I think it's a feature.

> there are no stopwords. We deal with mpl command either by skipping them/the snippet, or use IGNORE_OUTPUT, or use 
> the plot directive (then it's being tested during docs generation).

In the original scipy implementation an ability to skip MPL stuff was important. How do you skip/ignore MPL commands? 

refuide-check does 
https://github.com/scipy/scipy/blob/main/tools/refguide_check.py#L559 and https://github.com/scipy/scipy/blob/main/tools/refguide_check.py#L579

> new release of pytest regularly breaks doctestplus. I don't think that is doctestplus related, and feel that other plugins would be affected as well.

This is one nice feature of a bespoke solution like refguide-check : nearly zero maintenance over long time, no need to follow whatever pytest is doing.

> doctest likes to have full examples. E.g. import statements, and definitions. e.g. one cannot just say in the middle of a docstring that np.mean(input_array) without saying anything about np and input_array. But I think this is also a case with vanilla doctest.

Agreed, this is a doctest thing. What refuide-check does, it requires a numpy import and recognizes some numpy reprs (https://github.com/scipy/scipy/blob/main/tools/refguide_check.py#L477 and https://github.com/scipy/scipy/blob/main/tools/refguide_check.py#L606)


> `# doctest: +SKIP` and other comments are visible in the docstring from the python interpreter. Which is ugly. But they are rendered out by sphinx so the narrative docs/API docs will look great.

Which is precisely why additional `# may vary` markers come in handy.

> there are bugs but not enough maintainer time. So most of us fix the ones that bite our own projects. Thus a few, like that 
> whitespace+float bug is still around. But I think this is not a unique situation, and one of the main reasons why I feel it would 
> be great to pull resources together.

For whitespace I seriously suggest to consider the refguide-check implementation of the checker. This way the whole whitespace thing disappears in a puff of smoke :-).

If there's interest, I might look into contributing this.

**Comment by asmeurer** [comment]
Sorry if I missed it in the above discussion, but just how *are* you supposed to run the doctests in NumPy? I made a couple of edits to some docstrings and I'd like to make sure I didn't typo anything.

**Comment by bsipocz** [comment]
@asmeurer - I believe `python runtests.py --doctests -t <file_to_test>` should work.

I'm planning to finish the work we started at the NF summit over the holidays, and to have an obvious and consistent check for everything either way we choose to go.

**Comment by bsipocz** [comment]
Yes, I got long error messages, and don't understand why it's happening (haven't dug deep enough in the not too standard ways of test collection). 

And don't see why nothing was collected, this will be a good file for me to experiment with doctestplus.

**Comment by asmeurer** [comment]
I found `tools/refguide_check.py --doctests`, but it doesn't seem to run the doctest in question (I edited it to have a blatant error and it didn't fail).

Are the doctests actually run on CI anywhere? I didn't see anything from grepping, other than this refguide_check script, which is never run with the `--doctests` flag anyway.

**Comment by asmeurer** [comment]
Also presumably it should just be `./runtests.py --doctests numpy/linalg/linalg.py`? Maybe I'm misreading `./runtests.py --help` but it looks like `-t` is supposed to work like the `-k` flag to pytest. Anyway, this usage of runtests.py has never worked for me (e.g., I can't even get `./runtests.py numpy/linalg/tests` to work correctly). It's not clear to me if `runtests.py` is supposed to be functional or if you're just supposed to use `pytest` directly.

**Comment by asmeurer** [comment]
Oh I think the problem is that the doctester cannot see the functions defined in `numpy/linalg/linalg.py` as being defined there because of the way the decorators work. If you just use `numpy/linalg` instead it works, although it doesn't seem to know about numerical or whitespace differences. Also the doctest:+SKIP flag is not used in some places it should be. 

I'm coming to the conclusion that doctests just aren't run in NumPy. So I will just have be very careful with my doctest edits.

**Comment by bsipocz** [comment]
> Are the doctests actually run on CI anywhere?

I'm fairly certain that they are not, and my recollection of trying to force something to even run locally was clumsy due to the custom test collection. 

And good point about the decorations, they were one of the prime suspects for me, too why tests were not collected from that particular file.

**Comment by charris** [comment]
I think what you are looking for is `tools/refguide_check.py`, which can also be run with `runtests.py --refguide-check`. See the documentation of the tool. We run that for mac on azure-pipelines and also on circleci. And example given is `python tools/refguide_check.py --doctests ma`. Note that the tests pickup the installed NumPy, so the runtests version is probably best as it has a local environment (which doesn't work for me because I have so many installed versions and it never picks the right one ;) But maybe the tools version works, I've never tried it alone.

**Comment by rossbar** [comment]
> Are the doctests actually run on CI anywhere?

They are in the doc build on circleci: https://github.com/numpy/numpy/blob/a24ddad5d1d43ec9bea474abf8068ee6b8db01db/.circleci/config.yml#L72-L77

Though in practice I don't know if I've ever seen this job fail. This dovetails into the original point about having confidence that `refguide_check` is configured correctly/discovering doctests comprehensively.

**Comment by charris** [comment]
> I don't know if I've ever seen this job fail.

I have definitely seen the macos test fail, and pretty sure I've seen that with circleci also.

**Comment by ev-br** [comment]
The current way to run run modified doctests is `$ python runtests --refguide-check -s <submodule>`. https://github.com/numpy/numpy/pull/21962 suggets the way of making it more flexible, but this does not seem to be wanted by the numpy team :-).

> This dovetails into the original point about having confidence that refguide_check is configured correctly/discovering doctests comprehensively.

Can't speak of the way the CI is done in numpy, but the command-line usage certainly does what it means to do: discovers doctests in docstrings of public objects. The above PR has a switch for using the regular python doctest package discovery rules.

**Comment by pllim** [comment]
Re: pytest-doctestplus

If you are still interested in the package, please see https://github.com/astropy/pytest-doctestplus/issues/199 . Thank you!
